python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,520 @@
+        return None
+
+    def can_transform(self, target: Type) -> bool:
+        return True
+
+    def get_result_type(self, source: Type) -> Type:
+        return source
+
+    def preserve_order(self) -> bool:
+        return False
+
+    def satisfy_order(self, other) -> bool:
+        return self == other
+
+    def to_human_string(self, value) -> str:
+        return "null"
+
+    def dedup_name(self) -> str:
+        return self.__str__()
+
+
+class Bucket(Transform):
+    _MURMUR3 = mmh3
+    _MAX_32_BITS_INT = 2147483647
+    _FUNCTIONS_MAP = {  # [0] is hash function and [1] is can_transform check function
+        DateType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("q", v)),
+            lambda t: t in [IntegerType, DateType],
+        ),
+        IntegerType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("q", v)),
+            lambda t: t in [IntegerType, DateType],
+        ),
+        TimeType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("q", v)),
+            lambda t: t in {LongType, TimeType, TimestampType, TimestamptzType},
+        ),
+        TimestampType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("q", v)),
+            lambda t: t in {LongType, TimeType, TimestampType, TimestamptzType},
+        ),
+        TimestamptzType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("q", v)),
+            lambda t: t in {LongType, TimeType, TimestampType, TimestamptzType},
+        ),
+        LongType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("q", v)),
+            lambda t: t in [LongType, TimeType, TimestampType, TimestamptzType],
+        ),
+        StringType: (
+            lambda v: Bucket._MURMUR3.hash(v),
+            lambda t: t == StringType,
+        ),
+        BinaryType: (
+            lambda v: Bucket._MURMUR3.hash(v),
+            lambda t: t == BinaryType or isinstance(t, FixedType),
+        ),
+        UUIDType: (
+            lambda v: Bucket._MURMUR3.hash(
+                struct.pack(
+                    ">QQ",
+                    (v.int >> 64) & 0xFFFFFFFFFFFFFFFF,
+                    v.int & 0xFFFFFFFFFFFFFFFF,
+                )
+            ),
+            lambda t: t == UUIDType,
+        ),
+        # bucketing by Float/Double is not allowed by the spec, but they have hash implementation
+        FloatType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("d", v)),
+            lambda t: t == FloatType,
+        ),
+        DoubleType: (
+            lambda v: Bucket._MURMUR3.hash(struct.pack("d", v)),
+            lambda t: t == DoubleType,
+        ),
+    }
+
+    @staticmethod
+    def _decimal_to_bytes(value: Decimal):
+        unscaled_value = Transform._unscale_decimal(value)
+        number_of_bytes = int(math.ceil(unscaled_value.bit_length() / 8))
+        return unscaled_value.to_bytes(length=number_of_bytes, byteorder="big")
+
+    def __init__(self, transform_type: Type, num_buckets: int):
```

Comment:
I don't think that `transform_type` is accurate. That should be the source type, right?

Reply:
I saw that `source_type` is also used in `get_result_type` so I used `transform_type` to indicate this is the type for the transform. Or should we rename `get_result_type` using another arg name?

Reply:
replaced `transform_type` by `source_type`.